gpu4pyscf/tdscf/ris.py

```diff
@@ -0,0 +1,1653 @@
+    (ib|ja) = Σ_Pjb (T_ib^P T_ja^P V_jb^m)
+            = Σ_P [T_ja^P Σ_jb(T_ib^P V_jb^m)]           
+    '''
+    # def ibja_MVP(V):
+    #     T_ib_V = einsum("Pib,mjb->Pimj", T_ia, V)
+    #     ibja_V = einsum("Pja,Pimj->mia", T_ia, T_ib_V)
+    #     return ibja_V
+
+    def ibja_MVP(V, occ_chunk_size=100):
+        '''
+        Optimized calculation of (ib|ja) = Σ_Pjb (T_ib^P T_ja^P V_jb^m)
+        by chunking along the n_occ dimension to reduce memory usage.
+
+        Parameters:
+            V (cupy.ndarray): Input tensor of shape (n_state, n_occ, n_vir).
+            occ_chunk_size (int): Chunk size for splitting the n_occ dimension.
+
+        Returns:
+            ibja_V (cupy.ndarray): Result tensor of shape (n_state, n_occ, n_vir).
+        '''
+        nauxao, n_occ, n_vir = T_ia.shape
+        n_state, n_occ, n_vir = V.shape
+        # assert n_occ == n_occ_v and n_vir == n_vir_v, "Shapes of V and T_ia must match"
+
+        # Initialize result tensor
+        ibja_V = cp.empty((n_state, n_occ, n_vir), dtype=T_ia.dtype)
+
+        # Iterate over chunks of the n_occ dimension
+        for occ_start in range(0, n_occ, occ_chunk_size):
+            occ_end = min(occ_start + occ_chunk_size, n_occ)
+            #occ_range = occ_end - occ_start
+
+            # Extract the current chunk of V
+            V_chunk = V[:, occ_start:occ_end, :]  # Shape: (n_state, occ_range, n_vir)
+
+            # Extract the corresponding chunk of T_ia
+            T_ia_chunk = T_ia[:, occ_start:occ_end, :]  # Shape: (nauxao, occ_range, n_vir)
+
+            # Compute T_ib_V for the current chunk
+            T_ib_V_chunk = cp.einsum("Pib,mjb->Pimj", T_ia_chunk, V_chunk, optimize=True)
+
+            # Compute ibja_V for the current chunk
+            ibja_V[:, occ_start:occ_end, :] = cp.einsum("Pja,Pimj->mia", T_ia_chunk, T_ib_V_chunk)
+
+            # Release intermediate variables and clean up memory
+            # del V_chunk, T_ia_chunk, T_ib_V_chunk
+            # cp.get_default_memory_pool().free_all_blocks()
+
+        return ibja_V
+
+    return ibja_MVP
+
+class RisBase(lib.StreamObject):
+    def __init__(self, 
+                mf, 
+                theta: float = 0.2,
+                J_fit: str = 'sp',
+                K_fit: str = 's',
+                Ktrunc: float = 40.0,
+                a_x: float = None,
+                omega: float = None,
+                alpha: float = None,
+                beta: float = None,
+                conv_tol: float = 1e-3,
+                nstates: int = 5,
+                max_iter: int = 25,
+                spectra: bool = False,
+                out_name: str = '',
+                print_threshold: float = 0.05,
+                GS: bool = False,
+                single: bool = True,
+                group_size: int = 256,
+                group_size_aux: int = 256):
+
+        self.single = single
+        if single:
+            mf.mo_coeff = cp.asarray(mf.mo_coeff, dtype=cp.float32)

```

Comment:
don't overwrite mf.mo_coeff. this can easily cause potential bugs in other modules. You can create a mo_coeff for this instance and refer to the local mo_coeff attributes in your code. If the mf.mo_coeff must be overwritten, please make a copy `mf = mf.copy()` first